File: app/src/main/java/com/taitsmith/busboy/utils/NearbyAdapter.java
package com.taitsmith.busboy.utils;

import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.BaseAdapter;

import com.taitsmith.busboy.databinding.ListItemNearbyBinding;
import com.taitsmith.busboy.obj.Stop;

import java.util.HashMap;
import java.util.List;

public class NearbyAdapter extends BaseAdapter {
    HashMap<String, String> destinationList;
    List<Stop> stopList;
    ListItemNearbyBinding binding;

    public NearbyAdapter(HashMap<String, String> destinationList,
                         List<Stop> stopList) {
        this.destinationList = destinationList;
        this.stopList = stopList;
    }

    @Override
    public int getCount() {
        return stopList.size();
    }

    @Override
    public Object getItem(int i) {
        return null;
    }

    @Override
    public long getItemId(int i) {
        return i;
    }

    @Override
    public int getItemViewType(int position) {
        return position;
    }

    @Override
    public int getViewTypeCount() {
        if (stopList.size() == 0) {
            return 1;
        }
        return stopList.size();
    }

    @Override
    public View getView(int position, View view, ViewGroup parent) {
        ViewHolder holder;
        Stop stop = stopList.get(position);
        if (view == null) {
            binding = ListItemNearbyBinding.inflate(
                    LayoutInflater.from(parent.getContext()),
                    parent, false);

            holder = new ViewHolder(binding);
            holder.view = binding.getRoot();
            holder.view.setTag(holder);

        } else {
            holder = (ViewHolder) view.getTag();
        }

        binding.setStop(stop);


        binding.listItemLinesServed.setText(destinationList.get(stop.getStopId()));

        return holder.view;
    }

    private static class ViewHolder {
        private View view;
        ListItemNearbyBinding binding;

        ViewHolder(ListItemNearbyBinding binding) {
            this.view = binding.getRoot();
            this.binding = binding;
        }
    }
}

File: app/src/main/java/com/taitsmith/busboy/utils/NearbyAdapter.kt
package com.taitsmith.busboy.utils

import android.view.LayoutInflater
import android.view.ViewGroup
import androidx.recyclerview.widget.DiffUtil
import androidx.recyclerview.widget.ListAdapter
import androidx.recyclerview.widget.RecyclerView
import com.taitsmith.busboy.data.Stop
import com.taitsmith.busboy.databinding.ListItemNearbyBinding

class NearbyAdapter(
    private val onItemClicked: (Stop) -> Unit,
    private val onItemLongClick: (Stop) -> Unit
): ListAdapter<Stop, NearbyAdapter.NearbyViewHolder>(DiffCallback) {

    companion object {
        private val DiffCallback = object: DiffUtil.ItemCallback<Stop>() {
            override fun areItemsTheSame(oldItem: Stop, newItem: Stop): Boolean {
                return oldItem.stopId == newItem.stopId
            }

            override fun areContentsTheSame(oldItem: Stop, newItem: Stop): Boolean {
                return oldItem == newItem
            }
        }
    }

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): NearbyViewHolder {
        val viewHolder = NearbyViewHolder(
            ListItemNearbyBinding.inflate(
                LayoutInflater.from(parent.context),
                parent,
                false
            )
        )

        viewHolder.itemView.setOnClickListener {
            val position = viewHolder.bindingAdapterPosition
            onItemClicked(getItem(position))
        }

        viewHolder.itemView.setOnLongClickListener {
            val position = viewHolder.bindingAdapterPosition
            onItemLongClick(getItem(position))
            true
        }
        return viewHolder
    }

    override fun onBindViewHolder(holder: NearbyViewHolder, position: Int) {
        holder.bind(getItem(position))
    }

    class NearbyViewHolder(
        private var binding: ListItemNearbyBinding
    ): RecyclerView.ViewHolder(binding.root) {
        fun bind(stop: Stop) {
            binding.stop = stop
        }
    }
}